src/codegen/codegen_llvm.cpp

```diff
@@ -320,67 +333,265 @@ void CodeGen_LLVM::visit(const Rem* op) {
+    value = Builder->CreateMinNum(codegen(op->operands[0]), codegen(op->operands[1]));
+    for (size_t i = 2; i < op->operands.size(); i++) {
+      value = Builder->CreateMinNum(value, codegen(op->operands[i]));
+    }
+  } else {
+    auto a = codegen(op->operands[0]);
+    auto b = codegen(op->operands[1]);
+
+    llvm::Value* icmp =
+        (op->type.isInt()) ? Builder->CreateICmpSLT(a, b) : Builder->CreateICmpULT(a, b);
+    value = Builder->CreateSelect(icmp, a, b);
+
+    for (size_t i = 2; i < op->operands.size(); i++) {
+      auto c = codegen(op->operands[i]);
+      icmp =
+          (op->type.isInt()) ? Builder->CreateICmpSLT(value, c) : Builder->CreateICmpULT(value, c);
+      value = Builder->CreateSelect(icmp, value, c);
+    }
+  }
 }
 
 void CodeGen_LLVM::visit(const Max* op) {
   auto _ = CodeGen_LLVM::IndentHelper(this, "Max");
-  throw logic_error("Not Implemented for Max.");
+
+  if (op->operands.size() == 1) {
+    value = codegen(op->operands[0]);
+    return;
+  }
+
+  if (op->type.isFloat()) {
+    // LLVM's minnum intrinsic only does binary ops
+    value = Builder->CreateMaxNum(codegen(op->operands[0]), codegen(op->operands[1]));
+    for (size_t i = 2; i < op->operands.size(); i++) {
+      value = Builder->CreateMaxNum(value, codegen(op->operands[i]));
+    }
+  } else {
+    auto a = codegen(op->operands[0]);
+    auto b = codegen(op->operands[1]);
+
+    llvm::Value* icmp =
+        (op->type.isInt()) ? Builder->CreateICmpSGT(a, b) : Builder->CreateICmpUGT(a, b);
+    value = Builder->CreateSelect(icmp, a, b);
+
+    for (size_t i = 2; i < op->operands.size(); i++) {
+      auto c = codegen(op->operands[i]);
+      icmp =
+          (op->type.isInt()) ? Builder->CreateICmpSGT(value, c) : Builder->CreateICmpUGT(value, c);
+      value = Builder->CreateSelect(icmp, value, c);
+    }
+  }
 }
 
 void CodeGen_LLVM::visit(const BitAnd* op) {
   auto _ = CodeGen_LLVM::IndentHelper(this, "BitAnd");
-  throw logic_error("Not Implemented for BitAnd.");
+  value = Builder->CreateAnd(codegen(op->a), codegen(op->b));
 }
 
 void CodeGen_LLVM::visit(const BitOr* op) {
   auto _ = CodeGen_LLVM::IndentHelper(this, "BitOr");
-  throw logic_error("Not Implemented for BitOr.");
+  value = Builder->CreateOr(codegen(op->a), codegen(op->b));
 }
 
 void CodeGen_LLVM::visit(const Eq* op) {
   auto _ = CodeGen_LLVM::IndentHelper(this, "Eq");
-  throw logic_error("Not Implemented for Eq.");
+  auto* a = codegen(op->a);
+  auto* b = codegen(op->b);
+
+  if (op->type.isFloat()) {
+    if (auto* FPa = llvm::dyn_cast<llvm::ConstantFP>(a)) {
+      if (auto* FPb = llvm::dyn_cast<llvm::ConstantFP>(b)) {
+        if (FPa->isNaN() || FPb->isNaN())
+          value = Builder->CreateFCmpUEQ(a, b);  // ULE means that either operand may be a QNAN.
+        else
+          value = Builder->CreateFCmpOEQ(a, b);  // OLE means that neither operand is a QNAN
+      }
+    }
+  } else {
+    value = Builder->CreateICmpEQ(a, b);
+  }
```

Comment:
Change the else statement to check if `op->type.isInt() || op->type.isUInt()` is true
```suggestion
  } else if (op->type.isInt() || op->type.isUInt()) {
    value = Builder->CreateICmpEQ(a, b);
  } else {
     // raise an error here
  }
```

Reply:
Done!